xarray/core/combine.py

```diff
@@ -369,34 +370,178 @@ def _auto_concat(datasets, dim=None, data_vars='all', coords='different'):
+    # TODO create custom exception class instead of using asserts?
+    # Is this function even necessary?
+
+    tile_ids = combined_tile_ids.keys()
+
+    # Check all tuples are the same length
+    lengths = [len(id) for id in tile_ids]
+    assert set(lengths) == {lengths[0]}
+
+    # Check only datasets are contained
+    from .dataset import Dataset
+    for v in combined_tile_ids.values():
+        assert isinstance(v, Dataset)
+
+
+def _data_vars(combined_id):
+    id, ds = combined_id
+    return tuple(sorted(ds.data_vars))
+
+
+def _combine_nd(combined_IDs, concat_dims, data_vars='all',
+                coords='different', compat='no_conflicts'):
+    """
+    Concatenates and merges an N-dimensional structure of datasets.
+
+    No checks are performed on the consistency of the datasets, concat_dims or
+    tile_IDs, because it is assumed that this has already been done.
+
+    Parameters
+    ----------
+    combined_IDs : Dict[Tuple[int, ...]], xarray.Dataset]
+        Structure containing all datasets to be concatenated with "tile_IDs" as
+        keys, which specify position within the desired final combined result.
+    concat_dims : sequence of str
+        The dimensions along which the datasets should be concatenated. Must be
+        in order, and the length must match
+
+    Returns
+    -------
+
+    """
+
+    # Organise by data variables
+    grouped_by_data_vars = itertoolz.groupby(_data_vars,
+                                             combined_IDs.items()).values()
+
+    concatenated_datasets = []
+    for tiled_datasets_group in grouped_by_data_vars:
+
+        # Convert list of tuples back into a dictionary
+        concatenated_ids = dict(tiled_datasets_group)
+
+        # Perform N-D dimensional concatenation
+        # Each iteration of this loop reduces the length of the tile_IDs tuples
+        # by one. It always removes the first
+        for concat_dim in concat_dims:
+            dim = None if concat_dim is _CONCAT_DIM_DEFAULT else concat_dim
+
+            concatenated_ids = _concat_along_first_dim(concatenated_ids,
+                                                       dim=dim,
+                                                       data_vars=data_vars,
+                                                       coords=coords)
+        concatenated_datasets = concatenated_datasets \
+                                + list(concatenated_ids.values())
+    return merge(concatenated_datasets, compat=compat)
+
+
+def _new_tile_id(single_id_ds_pair):
+    # TODO maybe replace with something like lambda x: x[0][1:]?
+    tile_id, ds = single_id_ds_pair
+    return tile_id[1:]
+
+
+def _concat_along_first_dim(combined_IDs, dim, data_vars='all',
+                                coords='different'):
+    grouped = itertoolz.groupby(_new_tile_id, combined_IDs.items())
+    new_combined_IDs = {}
+
+    # TODO Would there be any point in parallelizing this concatenation step?
```

Comment:
Unlikely -- with dask, we parallelize processing the actual array data, but all this Python logic is hard to parallelize (due to Python's GIL). Unless you're dealing with a lot of very small datasets this wouldn't make a difference (and lot of small datasets will be slow regardless using Python).